Code/GraphMol/MarvinParse/MarvinDefs.cpp

```diff
@@ -0,0 +1,2978 @@
+  sgMap.clear();
+  atomMap.clear();
+  bondMap.clear();
+
+  for (auto &reactant : reactants) {
+    reactant->prepSgroupsForRDKit();
+    // clean and renumber the sgroups
+    reactant->clearMaps();
+    reactant->cleanUpNumbering(molCount, atomCount, bondCount, sgCount);
+  }
+  for (auto &agent : agents) {
+    agent->prepSgroupsForRDKit();
+    agent->cleanUpNumbering(molCount, atomCount, bondCount, sgCount);
+  }
+  for (auto &product : products) {
+    product->prepSgroupsForRDKit();
+    product->cleanUpNumbering(molCount, atomCount, bondCount, sgCount);
+  }
+}
+
+std::string MarvinReaction::toString() {
+  std::ostringstream out;
+
+  out << "<cml xmlns=\"http://www.chemaxon.com\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+         " xsi:schemaLocation=\"http://www.chemaxon.com http://www.chemaxon.com/marvin/schema/mrvSchema_20_20_0.xsd\">"
+         "<MDocument><MChemicalStruct><reaction>";
+
+  out << "<reactantList>";
+  for (auto &reactant : reactants) {
+    out << reactant->toString();
+  }
+  out << "</reactantList>";
+  out << "<agentList>";
+  for (auto &agent : agents) {
+    out << agent->toString();
+  }
+  out << "</agentList>";
+  out << "<productList>";
+  for (auto &product : products) {
+    out << product->toString();
+  }
+  out << "</productList>";
+
+  out << arrow.toString();
+
+  out << "</reaction></MChemicalStruct>";
+
+  for (auto &pluse : pluses) {
+    out << pluse->toString();
+  }
+  for (auto &condition : conditions) {
+    out << condition->toString();
+  }
+
+  out << "</MDocument></cml>";
+  return out.str();
+}
+
+ptree MarvinReaction::toPtree() const {
+  ptree out;
+
+  out.put("cml.<xmlattr>.xmlns", "http://www.chemaxon.com");
+  out.put("cml.<xmlattr>.xmlns:xsi",
+          "http://www.w3.org/2001/XMLSchema-instance");
+  out.put(
+      "cml.<xmlattr>.xsi:schemaLocation",
+      "http://www.chemaxon.com http://www.chemaxon.com/marvin/schema/mrvSchema_20_20_0.xsd");
+
+  for (auto &reactant : reactants) {
+    out.add_child(
+        "cml.MDocument.MChemicalStruct.reaction.reactantList.molecule",
+        reactant->toPtree());
+  }
+
+  for (auto &agent : agents) {
+    out.add_child("cml.MDocument.MChemicalStruct.reaction.agentList.molecule",
+                  agent->toPtree());
+  }
+
+  for (auto &product : products) {
+    out.add_child("cml.MDocument.MChemicalStruct.reaction.productList.molecule",
+                  product->toPtree());
+  }
+
+  out.put_child("cml.MDocument.MChemicalStruct.reaction.arrow",
+                arrow.toPtree());
+
+  for (auto &pluse : pluses) {
+    out.add_child("cml.MDocument.MReactionSign", pluse->toPtree());
+  }
+
+  for (auto &condition : conditions) {
+    out.add_child("cml.MDocument.MTextBox", condition->toPtree());
+  }
+
+  return out;
+}
+
+MarvinRectangle::MarvinRectangle() {
+  upperLeft.x = 0.0;
+  upperLeft.y = 0.0;
+  lowerRight.x = 0.0;
+  lowerRight.y = 0.0;
+  centerIsStale = true;
```

Comment:
this wouldn't be necessary if you just iniitalized the values in the class definition

Reply:
I really prefer to have each contructor do the initialization - there are several contructors and they are all different

Reply:
> I really prefer to have each constructor do the initialization - there are several constructors and they are all different

That's clear, but this particular one is just initializing values which should be handled in the class definition.

We're trying to get the code consistent and start adopting some common coding guidelines, but until we get those written down and applied more broadly across the code base, it's not worth arguing about
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-default

Reply:
Not worth arguing about - so I changed it to init the centerIsStale boolean.  As it turns out, the default constructor is never called anyway :)